src/config.rs

```diff
@@ -0,0 +1,211 @@
+/// The available configuration keys.
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, Display, EnumString, AsRefStr, EnumIter, EnumProperty,
+)]
+#[strum(serialize_all = "snake_case")]
+pub enum Config {
+    Addr,
+    MailServer,
+    MailUser,
+    MailPw,
+    MailPort,
+    SendServer,
+    SendUser,
+    SendPw,
+    SendPort,
+    ServerFlags,
+    #[strum(props(default = "INBOX"))]
+    ImapFolder,
+    Displayname,
+    Selfstatus,
+    Selfavatar,
+    #[strum(props(default = "1"))]
+    E2eeEnabled,
+    #[strum(props(default = "1"))]
+    MdnsEnabled,
+    InboxWatch,
+    #[strum(props(default = "1"))]
+    SentboxWatch,
+    #[strum(props(default = "1"))]
+    MvboxWatch,
+    #[strum(props(default = "1"))]
+    MvboxMove,
+    #[strum(props(default = "0"))]
+    ShowEmails,
+    SaveMimeHeaders,
+    ConfiguredAddr,
+    ConfiguredMailServer,
+    ConfiguredMailUser,
+    ConfiguredMailPw,
+    ConfiguredMailPort,
+    ConfiguredSendServer,
+    ConfiguredSendUser,
+    ConfiguredSendPw,
+    ConfiguredSendPort,
+    ConfiguredServerFlags,
+    Configured,
+}
+
+// deprecated
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Display, EnumString, AsRefStr, EnumIter)]
+pub enum SysConfig {
+    #[strum(serialize = "sys.version")]
+    Version,
+    #[strum(serialize = "sys.msgsize_max_recommended")]
+    MsgsizeMaxRecommended,
+    #[strum(serialize = "sys.config_keys")]
+    ConfigKeys,
+}
+
+/// Get a configuration key.
+/// Returns "" when the key is invalid, or no default was found.
+pub fn get(context: &Context, key: impl AsRef<str>) -> String {
+    let key = key.as_ref();
+
+    if key.starts_with("sys.") {
+        return get_sys_config_str(key);
+    }
+
+    match Config::from_str(key) {
+        Ok(config_key) => {
+            let value = match config_key {
+                Config::Selfavatar => {
+                    let rel_path = context.sql.get_config(context, key, None);
+                    rel_path.map(|p| {
+                        let v = unsafe { dc_get_abs_path(context, to_cstring(p).as_ptr()) };
+                        let r = to_string(v);
+                        unsafe { free(v as *mut _) };
+                        r
+                    })
+                }
+                _ => context.sql.get_config(context, key, None),
+            };
+
+            if value.is_some() {
+                return value.unwrap();
+            }
+
+            // Default values
+            match config_key {
+                Config::Selfstatus => {
+                    let s = unsafe { dc_stock_str(context, 13) };
+                    let res = to_string(s);
+                    unsafe { free(s as *mut _) };
+                    res
+                }
+                _ => config_key
+                    .get_str("default")
+                    .unwrap_or_default()
+                    .to_string(),
+            }
+        }
+        Err(_) => "".into(),
+    }
+}
+
+fn get_sys_config_str(key: impl AsRef<str>) -> String {
```

Comment:
I am kind of a fan of every function having a good docstring... While one might argue the code says the same it's often a good indicator when explaining it in english is complicated that the function could use some refactoring.

Reply:
you are not wrong, but I am lazy sometimes..